components/espressif/esp32-wroom-32.stanza

```diff
@@ -0,0 +1,277 @@
+  module(memory, pcb-antenna?, false, serial-programming?) 
+
+public pcb-module module (memory:Int, pcb-antenna?:True|False, high-temp:True|False, serial-programming?:True|False) :
+  public inst esp32 : ocdb/espressif/esp32-wroom-32/component(memory, pcb-antenna?, high-temp)
+  port power : power
+  pin boot
+  pin en
+
+; RC enable delay
+  inst en-r: chip-resistor(10.0e3)
+  inst en-c: ceramic-cap(1.0e-6)
+  net (power.gnd esp32.GND en-c.p[1])
+  net (power.vdd esp32.P3V3 en-r.p[1])
+  net (esp32.EN en-c.p[2] en-r.p[2])
+
+  ; supports power:
+  ;   power.vdd => esp32.P3V3
+  ;   power.gnd => esp32.GND
+  
+  supports reset:
+    reset.reset => esp32.EN
+
+  supports jtag([]) :
+    jtag([]).tms => esp32.IO[14]
+    jtag([]).tdi => esp32.IO[12]
+    jtag([]).tck => esp32.IO[13]
+    jtag([]).tdo => esp32.IO[15]
+
+  net (esp32.IO[0], boot)
+  net (esp32.EN en)
+  if serial-programming? :
+    port uart : uart([UART-DTR UART-RTS UART-RX UART-TX])
+    net ESP-RX (esp32.RXD0, uart.rx)
+    net ESP-TX (esp32.TXD0, uart.tx)
+    public inst q : ocdb/nexperia/PMBT3904/component[2]
+    res-strap(uart.dtr, q[0].b, 10.0e3)
+    res-strap(uart.rts, q[1].b, 10.0e3)
+    net ( q[1].c boot)
+    net ( q[0].c en)
+    net (uart.rts, q[0].e)
+    net (uart.dtr, q[1].e)
+
+  for i in 0 to 22 do:
+    supports gpio:
+      require io:io-pin from esp32
+      gpio.gpio => io.p
+      property(io.p.digital-io) = DigitalIO(CMOSOutput(min-max(0.0, 0.1 * 3.3), min-max(0.8 * 3.3, 3.3)), 0.25 * 3.3, 0.75 * 3.3, esp32.P3V3, esp32.GND, 50.0e-9)
+  for i in 0 to 16 do:
+    supports timer:
+      require io:io-pin from esp32
+      timer.timer => io.p
+      property(io.p.digital-io) = DigitalIO(CMOSOutput(min-max(0.0, 0.1 * 3.3), min-max(0.8 * 3.3, 3.3)), 0.25 * 3.3, 0.75 * 3.3, esp32.P3V3, esp32.GND, 50.0e-9)
+  for i in 0 to 2 do:
+    supports i2c:
+      require io:io-pin[2] from esp32
+      i2c.sda => io[0].p
+      i2c.scl => io[1].p
+      property(io[0].p.digital-io) = DigitalIO(OpenCollector(min-max(0.0, 0.1 * 3.3), 28.0e-3), 0.25 * 3.3, 0.75 * 3.3, esp32.P3V3, esp32.GND, 50.0e-9)
+      property(io[1].p.digital-io) = DigitalIO(OpenCollector(min-max(0.0, 0.1 * 3.3), 28.0e-3), 0.25 * 3.3, 0.75 * 3.3, esp32.P3V3, esp32.GND, 50.0e-9)
+  for i in 0 to 2 do:
+    val ocdb-uart = /uart()
+    supports ocdb-uart:
+      require io:io-pin[2] from esp32
+      ocdb-uart.tx => io[0].p
+      property(io[0].p.digital-output) = DigitalOutput(CMOSOutput(min-max(0.0, 0.1 * 3.3), min-max(0.8 * 3.3, 3.3)), false, esp32.P3V3, esp32.GND)
+      ocdb-uart.rx => io[1].p
+      property(io[1].p.digital-input) = DigitalInput(0.25 * 3.3, 0.75 * 3.3, esp32.P3V3, esp32.GND, 50.0e-9)  
+
+  schematic-group(self) = ESP32
+
+; ================================================== Land Pattern ==================================================
+pcb-landpattern esp32-wroom (pcb-antenna?:True|False) :
```

Comment:
Is there a doc link for the package? Just curious if this is a weird one that the generators don't help with

Reply:
https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32e_esp32-wroom-32ue_datasheet_en.pdf
this is where I can use the most help

Reply:
Yea that's what I figured, it's a weird one that doesn't fit the existing generators well. I don't think we have a working custom LGA (which would be the closest to this example) that will work. I think reusing the internals is appropriate until we can design a more generic generator that can handle these cases, but it's not obvious if that's possible.